Add PMC to track events for newly enabled and disabled payment methods

hsingyuc opened the pull request:
Fixes #3920 

## Changes proposed in this Pull Request:
This PR adds the newly enabled and disabled payment methods track that being removed from #4091 back.

## Testing instructions
1. Install and active [WooCommerce Beta Tester 3.00](https://github.com/woocommerce/woocommerce/releases/download/wc-beta-tester-3.0.0/woocommerce-beta-tester.zip)
2. Go to Stripe settings, enable and disable payment methods
3. Go to status->logs `wp-admin/admin.php?page=wc-status&tab=logs`
4. Find the tracks file you just created
5. Check events `wcadmin_wcstripe_payment_method_settings_enabled` with payment methods you just enabled. (`wp-admin/admin.php?page=wc-status&tab=logs`)
6. Check events `wcadmin_wcstripe_payment_method_settings_disabled` with payment methods you just disabled

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [ ] Tested on mobile (or does not apply)

### Changelog entry

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

hsingyuc committed:
Track when payment methods are newly enabled or disabled in the PMC

hsingyuc committed:
Fix track newly disabled payment methods

hsingyuc committed:
Add changelog entry

wjrosa commented on changelog.txt:
```diff
@@ -90,6 +91,8 @@
 * Update - Update Amazon Pay icon to use image from WooCommerce Design Library.
 * Add - Show upcoming legacy checkout experience deprecation notice.
 * Fix - Fix variable dump in ACH customer error message when retrying a payment.
+* Dev - Fetch Stripe settings with Stripe configuration API.
```
Are all the entries intended?

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -646,6 +646,40 @@ public function get_upe_available_payment_methods() {
+	 * @param string[] $payment_method_ids_to_enable
+	 */
+	public function update_enabled_payment_methods( $payment_method_ids_to_enable ) {
+		WC_Stripe_Payment_Method_Configurations::update_payment_method_configuration( $payment_method_ids_to_enable, $this->get_stripe_supported_payment_methods() );
```
Nice refactor 👍

wjrosa commented on includes/admin/stripe-settings.php:
```diff
@@ -307,14 +307,6 @@
 			'desc_tip'    => true,
 		],
 	];
-	if ( WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
```
Just to confirm that this removal was intended.

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-method-amazon-pay.php:
```diff
@@ -73,7 +73,8 @@ public static function is_amazon_pay_enabled() {
 		}
 
-		$upe_enabled_method_ids = WC_Stripe_Helper::get_settings( null, 'upe_checkout_experience_accepted_payments' );
+		$wc_stripe_upe_payment_method_amazon_pay = new self();
```
Can't this still be a static call? Also, why does this instantiate `self` while the implementation below instantiates `WC_Stripe_UPE_Payment_Gateway`? I think it would be nice to have a standard.

wjrosa commented on includes/class-wc-stripe-payment-method-configurations.php:
```diff
@@ -0,0 +1,160 @@
+		}
+
+		foreach ( $payment_method_configurations as $payment_method_configuration ) {
+			if ( ! $payment_method_configuration->livemode && $payment_method_configuration->parent && 'pmc_1LEKjBGX8lmJQndTBOzjqxSa' === $payment_method_configuration->parent ) {
```
It would be nice to have these `pmc_` IDs as constants.

wjrosa commented on includes/class-wc-stripe-payment-method-configurations.php:
```diff
@@ -0,0 +1,160 @@
+
+		if ( $merchant_payment_method_configuration ) {
+			foreach ( $merchant_payment_method_configuration as $payment_method_id => $payment_method ) {
+				if ( is_object( $payment_method ) &&
```
You can do something like this instead:
```php
			foreach ( $merchant_payment_method_configuration as $payment_method_id => $payment_method ) {
				if ( 'on' === ( $payment_method->display_preference->value ?? null ) ) {
					$enabled_payment_method_ids[] = $payment_method_id;
				}
			}
```

wjrosa commented on woocommerce-gateway-stripe.php:
```diff
@@ -645,21 +646,24 @@ protected function enable_upe( $settings ) {
+						$payment_methods_to_enable[] = $method_class::STRIPE_ID;
 					}
 
 					if ( 'stripe' === $lpm_gateway_id && isset( $this->stripe_gateway ) && $this->stripe_gateway->is_enabled() ) {
```
This might be a good opportunity to use the constants here (and below).

hsingyuc replied:
Thank you, Wesley! I'll update this in another PR.

wjrosa commented:
The WooCommerce Beta Tester 3.00 link above (in the testing instructions) was not working for me. When I tried to install it, I got an error. I downloaded it from https://github.com/woocommerce/woocommerce/releases/tag/wc-beta-tester-3.0.0 instead.

wjrosa reviewed (changes requested):
Maybe I am doing something wrong, but I installed the extension (this version) on a new JN website, installed the Beta Tester extension, connected my Stripe account, and attempted to enable/disable methods on both the wp-admin settings page and the Stripe dashboard. None seemed to trigger the specific Track events mentioned (or logs since the instructions ask to take a look there).

hsingyuc committed:
Check if display_preference is set

hsingyuc commented:
Sorry, I updated the Beta tester link and rebased this branch.  I tested again, and below is how I tested it. Can you share what do you see?


https://github.com/user-attachments/assets/81f8abba-0b0c-4df3-b672-643df097349f

wjrosa commented on includes/class-wc-stripe-payment-method-configurations.php:
```diff
@@ -96,8 +96,20 @@ public static function get_upe_enabled_payment_method_ids() {
 		foreach ( $available_payment_method_ids as $stripe_id ) {
+			$will_enable = in_array( $stripe_id, $enabled_payment_method_ids, true );
+
+			if ( isset( $payment_method_configuration->$stripe_id->display_preference->value ) && 'on' === $payment_method_configuration->$stripe_id->display_preference->value && ! $will_enable ) {
```
Here, you can do the same as I mentioned before (replacing the `isset` use):
```php
'on' === ( $payment_method_configuration->$stripe_id->display_preference->value ?? null )
```

wjrosa commented:
> Sorry, I updated the Beta tester link and rebased this branch. I tested again, and below is how I tested it. Can you share what do you see?

Oh right! I was trying to look into the Tracks dashboard or in the Stripe logs, not in the `tracks` logs. I tried again and it worked:

<img width="739" alt="Screenshot 2025-04-02 at 12 19 44" src="https://github.com/user-attachments/assets/7a4a574a-e1cf-4af5-b6b3-80c533725de7" />

Left an optional suggestion 👍

hsingyuc committed:
Update for PR feedback

diegocurbelo committed:
Merge branch 'add/hook-up-enabled-payment-methods' into add/pmc-tracks

diegocurbelo committed:
Merge branch 'add/hook-up-enabled-payment-methods' into add/pmc-tracks
